### PR TITLE
Add EIP-2612 support for ERC20 (permits)

### PR DESCRIPTION
**Pull request by Lumyo**
Draft PR (Still need to write tests + signature generation function in lib)

### CODE REVIEW

**Comment by ItsNickBarry** [comment]
@Lumyo Merge the typescript branch into this branch if that helps with development, and we'll merge both into master when ready.

**Comment by Lumyo** [comment]
@ItsNickBarry its already branched from typescript branch with target branch of the PR to typescript (ill change the PR target to master once typescript branch is merged)

File: contracts/token/ERC20/ERC20Permit.sol [hunk 1]
```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "./ERC20.sol";
```
Comment (ItsNickBarry) [inline contracts/token/ERC20/ERC20Permit.sol:4]: Inherit from `ERC20Base`.

File: contracts/token/ERC20/ERC20Permit.sol [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+/**
+ * @dev Extension of {ERC20} that allows token holders to use their tokens
+ * without sending any transactions by setting {IERC20-allowance} with a
+ * signature using the {permit} method, and then spend them via
+ * {IERC20-transferFrom}.
+ *
+ * The {permit} signature mechanism conforms to the {IERC2612Permit} interface.
+ */
+abstract contract ERC20Permit is ERC20, IERC2612Permit {
+  constructor() {
```
Comment (ItsNickBarry) [inline contracts/token/ERC20/ERC20Permit.sol:19]: Avoid using constructors - instead put something in the test suite to validate that this is called in the end user's constructor or `initialize` function.  This is for proxy compatibility.

File: contracts/token/ERC20/ERC20PermitMock.sol [hunk 3]
```diff
@@ -0,0 +1,15 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import "./ERC20Permit.sol";
+
+contract ERC20ExtendedMock is ERC20Permit {
```
Comment (ItsNickBarry) [inline contracts/token/ERC20/ERC20PermitMock.sol:7]: Incorrect contract name.

File: contracts/token/ERC20/ERC20Permit.sol (hunk 1 above)
Comment (Lumyo) [reply contracts/token/ERC20/ERC20Permit.sol]: I had to also add inheritance from `ERC20Metadata`
Do we keep inheritance from both `ERC20Base` + `ERC20Metadata` or do I revert to only inheritance from `ERC20` ?

**Comment by Lumyo** [comment]
Tests require https://github.com/dmihal/eth-permit/pull/3 to be merged

File: contracts/token/ERC20/ERC20Permit.sol (hunk 1 above)
Comment (ItsNickBarry) [reply contracts/token/ERC20/ERC20Permit.sol]: Seems strange that `name()` is needed for this, but if that's the case then `ER20Base` + `ERC20Metadata` inheritance is correct.

**Comment by ItsNickBarry** [comment]
@Lumyo Merged for upcoming release, please check changes.  Failing tests can wait.

**Pull request merged**